File: algorithm/deque.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> v(20003);

int main(void) {
    ios_base::sync_with_stdio(false);
    int begin = 10000, end = 10001;
    int n; for (cin >> n; n--;) {
        string cmd; cin >> cmd;
        if (cmd == "push_front") {
            int num; cin >> num;
            v[begin--] = num;
        } else if (cmd == "push_back") {
            int num; cin >> num;
            v[end++] = num;
        } else if (cmd == "pop_front") {
            if (begin == end - 1) {
                cout << "-1" << "\n";
            } else {
                cout << v[++begin] << "\n";
                v[begin] = 0;
            }
        } else if (cmd == "pop_back") {
            if (begin == end - 1) {
                cout << "-1" << "\n";
            }
            else {
                cout << v[--end] << "\n";
                v[end] = 0;
            }
        } else if (cmd == "size") {
            cout << end - begin - 1 << "\n";
        } else if (cmd == "empty") {
            cout << (begin == end - 1 ? 1 : 0) << "\n";
        } else if (cmd == "front") {
            cout << (begin == end - 1 ? -1 : v[begin + 1]) << "\n";
        } else if (cmd == "back") {
            cout << (begin == end - 1 ? -1 : v[end - 1]) << "\n";
        }
    }
    return 0;
}

File: algorithm/deque.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class DequeExample {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int arr[] = new int[20003];
        int begin = 10000, end = 10001;
        int n = Integer.parseInt(br.readLine());
        for(int i = 0; i < n; ++i) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String cmd = st.nextToken();
            if(cmd.equals("push_front")) {
                int num = Integer.parseInt(st.nextToken());
                arr[begin--] = num;
            } else if(cmd.equals("push_back")) {
                int num = Integer.parseInt(st.nextToken());
                arr[end++] = num;
            } else if(cmd.equals("pop_front")) {
                if(begin == end - 1) {
                    sb.append("-1").append("\n");
                } else {
                    sb.append(arr[++begin]).append("\n");
                    arr[begin] = 0;
                }
            } else if(cmd.equals("pop_back")) {
                if(begin == end - 1) {
                    sb.append("-1").append("\n");
                } else {
                    sb.append(arr[--end]).append("\n");
                    arr[end] = 0;
                }
            } else if(cmd.equals("size")) {
                sb.append(end - begin - 1).append("\n");
            } else if(cmd.equals("empty")) {
                sb.append(begin == end - 1 ? 1 : 0).append("\n");
            } else if(cmd.equals("front")) {
                sb.append(begin == end - 1 ? -1 : arr[begin + 1]).append("\n");
            } else if(cmd.equals("back")) {
                sb.append(begin == end - 1? -1 : arr[end - 1]).append("\n");
            }
        }
        System.out.println(sb);
    }
}
